src/main/scala/mesosphere/marathon/core/deployment/impl/DeploymentActor.scala

```diff
@@ -155,17 +155,27 @@ private class DeploymentActor(
     async {
       val instances = await(instanceTracker.specInstances(runnableSpec.id))
       val runningInstances = instances.filter(_.state.condition.isActive)
-      val ScalingProposition(tasksToKill, tasksToStart) = ScalingProposition.propose(
+      val ScalingProposition(instancesToKill, tasksToStart) = ScalingProposition.propose(
         runningInstances, toKill, killToMeetConstraints, scaleTo, runnableSpec.killSelection)
 
-      def killTasksIfNeeded: Future[Done] = {
-        logger.debug("Kill tasks if needed")
-        tasksToKill.fold(Future.successful(Done)) { tasks =>
-          logger.debug("Kill tasks {}", tasks)
-          killService.killInstances(tasks, KillReason.DeploymentScaling).map(_ => Done)
+      def killTasksIfNeeded(): Future[Done] = {
+        def killAndChangeGoal(instances: Seq[Instance]): Future[Done] = async {
```

Comment:
why? it's really an implementation detail of this method and I needed the def only for async/await

Reply:
Testing. I just don't like nested methods so much. I guess it's because they rely on the closure capture instead of the method parameters. This makes them somewhat implicit:
```scala
def longMethod(some: String): String = {
  ...
  def bar(): String = {
     s"Where did $some come from?"
  }
  ...

  bar()
}
```

vs

```scala

def bar(input: String): String = {
  s"I'm pure and can be tested different $input"
}

def longMethod(some: String): String = {
  ...
  bar(some)
}
```

Reply:
To be fair your method has side effects and `killTasksIfNeeded()` takes no parameters.

Reply:
I would argue that you should not test private methods anyway, only public. And I really don't think this should be a public method.

But now I see that you are talking about `killTasksIfNeeded` while I did not introduce that method, I introduced killAndChangeGoal. But since now we are three levels deep :D I guess I an separate at least the `killTasksIfNeeded` - my `killAndChangeGoal` takes all the parameters it needs as a method should do :)